[Mobile] - Deprecate Gallery V1 as an Unsupported block

geriux opened the pull request:
## What?
Deprecates the V1 of the Gallery block since three versions of WordPress have been released since version 2 was released in `5.9`.

## Why?
To avoid maintaining old code that is no longer supported for too long.

## How?
Since there is still some usage we will use the current `__unstableGalleryWithImageBlocks` flag that was previously pulled from `gutenberg_experiments_editor_settings` but now it will be in the mobile block editor settings endpoint, to show users that have this flag as `false` an unsupported version of the Gallery block.

It will provide hints to upgrade to a newer version of WordPress and editing capability using a WebView in some specific cases.

Users that have this flag as `false` **won't see the Gallery block button in the inserter**. We could provide a hint every time they open the inserter or the editor but maybe it could be annoying for users that might not even use this block?

## Testing Instructions

**Precondition:**

- Use a self-hosted site
- Install the Gutenberg plugin and activate it
- Using the Plugin file editor, go to the Gutenberg plugin and update [this line](https://github.com/WordPress/gutenberg/blob/trunk/lib/experiments-page.php#L141-L143) to: 
 ```
	$experiments_settings = array(
		'__unstableGalleryWithImageBlocks' => false,
	);
```
- Using the web editor create a post with a Gallery block

On the mobile app:

- Log into your self-hosted site
- Open the draft post you created
- Expect to see the Gallery block as `Unsupported` if you don't see this, re-open the editor
- Tap on the `?` button
- Expect to see a description of why this block is unsupported

## Screenshots or screencast <!-- if applicable -->
<kbd><img src="https://user-images.githubusercontent.com/4885740/204553680-793a3937-b257-4a27-94b1-28a38831a727.png" width=250 /></kbd>

Commit:
Mobile - Deprecate Gallery V1 as an Unsupported block

Commit:
Add value fallback

Commit:
Mobile - Use custom title and description for specific unsupported blocks

Commit:
Add missing end of line

fluiddot commented on lib/experimental/block-editor-settings-mobile.php:
```diff
@@ -31,6 +31,8 @@ function gutenberg_get_block_editor_settings_mobile( $settings ) {
 		// To tell mobile that the site uses list v2 (inner blocks).
 		$settings['__experimentalEnableListBlockV2'] = true;
+		// To tell mobile if it should mark the Gallery block as Unsupported for older versions of WordPress.
+		$settings['__unstableGalleryWithImageBlocks'] = is_wp_version_compatible( '5.9' );
```
I'm wondering if we should also include the condition `(int) get_option( 'use_balanceTags' ) !== 1` as we have here 🤔 :
https://github.com/WordPress/gutenberg/blob/ae560c85c0235e2b1a16cc2ec580c9e1b449adba/lib/experiments-page.php#L142

geriux replied:
To be honest, I'm not really sure what that's doing but let me ping @glendaviesnz do you know by any chance what `(int) get_option( 'use_balanceTags' ) !== 1` does? And if that would be needed for mobile? I think with just having `is_wp_version_compatible( '5.9' )` would be enough but it'd be nice to confirm, thank you!

glendaviesnz replied:
In versions of WP < 5.9 if `use_balanceTags` was enabled the post HTML was parsed and rewritten to make sure all the HTML tags were correctly balanced, but the code doing this didn't accept `figure` as being nestable so would rewrite any v2 gallery code to remove the nesting. 

This was [updated in core here](https://github.com/WordPress/wordpress-develop/pull/1678) so that the gallery code was left nested as is even if `use_balanceTags` is enabled on a site. So, as long as a site is >= 5.9 there should be no issues even if `use_balanceTags` is enabled.

fluiddot commented on packages/block-library/src/missing/edit.native.js:
```diff
@@ -120,6 +127,26 @@ export class UnsupportedBlockEdit extends Component {
 	}
 
+	getSheetTitle( blockTitle ) {
+		if ( CUSTOM_UNSUPPORTED_BLOCK_MESSAGE[ blockTitle ] ) {
```
Maybe instead of using the block's title, we could use the name (i.e. `core/gallery`) as it might be closer to being an identifier and hence less subject to change, wdyt?

geriux replied:
Good idea, I updated it 👍

fluiddot reviewed (commented):
I think this approach is solid to continue working on it @geriux 🎊. I added a couple of comments but nothing critical. Thanks 🙇 !

Commit:
Mobile - Use block name instead of using the title as an indentifier

Commit:
Mobile - Add integration test for the Gallery V1 deprecation if the galleryWithImageBlocks flag is set to false

Commit:
Mobile - Unsupported block - Use blockName instead of blockTitle

geriux commented:
> I think this approach is solid to continue working on it @geriux 🎊. I added a couple of comments but nothing critical. Thanks 🙇 !

Thank you for checking it @fluiddot this is now ready for review and testing. I added a small integration test for the feature flag.

fluiddot commented on packages/block-library/src/missing/edit.native.js:
```diff
@@ -35,6 +35,13 @@ import styles from './style.scss';
 // Blocks that can't be edited through the Unsupported block editor identified by their name.
 const UBE_INCOMPATIBLE_BLOCKS = [ 'core/block' ];
 const I18N_BLOCK_SCHEMA_TITLE = 'block title';
+const CUSTOM_UNSUPPORTED_BLOCK_MESSAGE = {
+	'core/gallery': {
+		/* translators: Unsupported block alert title. %s: The localized block name */
+		title: __( "'%s' needs an updated version of WordPress" ),
+		description: __( 'Please update to version 5.9 or above' ),
+	},
+};
```
Probably an edge case but if for any other reason the Gallery block is missing, using this message could be misleading to the user.

In [this PR](https://github.com/WordPress/gutenberg/pull/46215/files#diff-f7e97c6adf52713aec986e0817c301fd3bb0d904074f105268f469f35483f5ecR154-R155) I explored passing the block name to the `native.missing_block_detail` filter hook, this way we could override the details string by block. We could use a similar approach here and, instead of setting the custom message here, use the filter hook in the [`galleryCheck` function](https://github.com/WordPress/gutenberg/pull/46157/files#diff-3286cc974baacc1ef30a6c7d6c18ae3ea400c86e1223c88029c1ea15051d65ecR201-R206) and based on the value of `galleryWithImageBlocks` override the string. For this approach, we'd also need to define a new filter hook to allow overriding the sheet title. WDYT?

fluiddot commented on packages/block-library/src/gallery/test/index.native.js:
```diff
@@ -618,3 +629,39 @@ describe( 'Gallery block', () => {
+		const unsupportedBlock = await getBlock( screen, 'Unsupported' );
+		expect( unsupportedBlock ).toBeVisible();
+
+		expect( getEditorHtml() ).toMatchSnapshot();
```
I'm wondering if we expect the HTML to be different, even though the block is unsupported the HTML output will remain the same, is this accurate? If so, we might not need this expectation.
